### PR TITLE
HLS playlistType event

### PR DESCRIPTION
**Pull request by rhummelmose**
Draft implementation as discussed in #1265

### CODE REVIEW

**Comment by rhummelmose** [comment]
@erankor I've verified that this works in HLS.js, but VLC for instance won't let me rewind. I'm thinking it might be a VLC limitation though: https://trac.videolan.org/vlc/ticket/15994

**Comment by rhummelmose** [comment]
@erankor This addition seems to do the trick to a large extent but a very important thing has to be changed. The internal caching behavior prevents the module from calling the upstream for getting the VOD map after the first time it has been resolved.
Looking into the implementation of the state machine that handles the meta data section I've been unable to identify a way to cleverly circumvent the cache, without adding a explicit conditional that I'm thinking must be there somewhere for handling of live already. It would be awesome if you could help out a bit.
Thanks!

**Comment by rhummelmose** [comment]
Realized now after playing a little bit more with it, that it just hits the response cache, which is expected as I had it turned on. Turning it off makes it work perfectly.

**Comment by rhummelmose** [comment]
@erankor Based on your feedback in #1297 I've implemented the above changes/additions.

Please have a look when time allows :)

**Comment by erankor** [comment]
This change is problematic... if you will search the code for `media_set->type` you will see all the places that it will probably break. For example - dash_packager_write_mpd_period has a switch on this value, if `media_set->type` is MEDIA_SET_EVENT, it won't print the period header.
It may be better to avoid adding a new media_set type and instead just add some flag that will make the module output the tag you want. You can still use the existing type field in the JSON to relay this info (I mean - it won't be a new property on the JSON), just enable the flag + set the type to MEDIA_SET_LIVE.
The main point is - 99% of the code doesn't care if it's an event, it just cares if it's live/vod, so I think it's better to avoid touching all these places.

**Comment by rhummelmose** [comment]
@erankor thanks, I'll revise

**Comment by rhummelmose** [comment]
@erankor please let me know if it wasn't exactly what you had in mind, thanks again :)

**Comment by rhummelmose** [comment]
@erankor friendly bump :) I have this branch running in production now and it works perfectly. Please have a look when you have the time.

File: vod/media_set_parser.c [hunk 1]
```diff
@@ -2865,12 +2870,16 @@ media_set_parse_json(
 				parse_all_clips = FALSE;
 			}
 
+			// applying live window with infinite duration is necessary for playlistType event
+			// because it ensures that incomplete segments are not included in the resulting playlist
```
Comment (erankor) [inline vod/media_set_parser.c:2874]: I don't think it's about incomplete segments, it's about the fact we are not allowed to remove segments from the head of the media playlist when the type is event

**Review by erankor — CHANGES_REQUESTED** [review]

File: vod/media_set.h [hunk 2]
```diff
@@ -124,6 +124,7 @@ typedef struct {
 	vod_str_t id;
 	uint32_t type;
 	uint32_t original_type;					// will contain live in case of a live playlist that was forced to vod
+	bool_t set_event_playlist_type;			// causes HLS playlist type to be event and infinite live_window_duration
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply vod/media_set.h:127]: Agreed

File: README.md [hunk 3]
```diff
@@ -436,7 +436,7 @@ Mandatory fields:
 	
 Optional fields:
 * `id` - a string that identifies the set. The id can be retrieved by `$vod_set_id`.
-* `playlistType` - string, can be set to `live` or `vod`, default is `vod`.
+* `playlistType` - string, can be set to `live`, `vod` or `event` (only supported or HLS playlists), default is `vod`.
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply README.md:439]: Thanks

File: vod/hls/m3u8_builder.c [hunk 4]
```diff
@@ -9,6 +9,7 @@
 // macros
 #define M3U8_HEADER_PART1 "#EXTM3U\n#EXT-X-TARGETDURATION:%uL\n#EXT-X-ALLOW-CACHE:YES\n"
 #define M3U8_HEADER_VOD "#EXT-X-PLAYLIST-TYPE:VOD\n"
+#define M3U8_HEADER_EVENT "#EXT-X-PLAYLIST-TYPE:EVENT\n"
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply vod/hls/m3u8_builder.c:12]: Yes, I've made it calculate that part of the result_size based on the playllist type

File: vod/media_set_parser.c [hunk 5]
```diff
@@ -2119,8 +2120,7 @@ media_set_apply_clip_from(
 		// Note: aligning to keyframes only in case of vod, since in live, 
 		//	alignment to keyframes will happen in segmenter_get_live_window
 		sequence = &media_set->sequences[0];
-		if (sequence->key_frame_durations != NULL && 
-			media_set->type == MEDIA_SET_VOD)
+		if (sequence->key_frame_durations != NULL && media_set->type == MEDIA_SET_VOD)
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply vod/media_set_parser.c:2123]: You're right

File: vod/media_set_parser.c [hunk 6]
```diff
@@ -2478,15 +2478,20 @@ media_set_parse_json(
 		return rc;
 	}
 
-	// vod / live
+	result->set_event_playlist_type = params[MEDIA_SET_PARAM_PLAYLIST_TYPE] != NULL &&
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply vod/media_set_parser.c:2481]: Makes sense to make the normal case of VOD circumvent the rest of the code

File: vod/media_set_parser.c [hunk 7]
```diff
@@ -2737,7 +2742,7 @@ media_set_parse_json(
 				}
 
 				// initialize the look ahead segment times
-				if (result->type == MEDIA_SET_LIVE &&
+				if ((result->type == MEDIA_SET_LIVE) &&
```
[the remark this answers is not in the document]
Comment (rhummelmose) [reply vod/media_set_parser.c:2745]: Yes sir

**Review by rhummelmose — COMMENTED** [review]
As requested

File: vod/hls/m3u8_builder.c [hunk 8]
```diff
@@ -467,9 +468,15 @@ m3u8_builder_build_index_playlist(
...
+	int playlist_type_header_size = media_set->type == MEDIA_SET_VOD ?
+		sizeof(M3U8_HEADER_VOD) :
+		media_set->is_live_event ?
+		sizeof(M3U8_HEADER_EVENT) :
+		0;
+
 	result_size =
 		sizeof(M3U8_HEADER_PART1) + VOD_INT64_LEN +
-		sizeof(M3U8_HEADER_VOD) +
+		playlist_type_header_size +
```
Comment (erankor) [inline vod/hls/m3u8_builder.c:479]: just use M3U8_HEADER_EVENT here, not trying to be exact (e.g. before the change, I always added vod, also in live)

File: vod/media_set_parser.c [hunk 9]
```diff
@@ -2485,9 +2486,21 @@ media_set_parse_json(
 	{
 		result->presentation_end = TRUE;
 	}
-	else if (params[MEDIA_SET_PARAM_PLAYLIST_TYPE]->v.str.len == playlist_type_live.len &&
-		vod_strncasecmp(params[MEDIA_SET_PARAM_PLAYLIST_TYPE]->v.str.data, playlist_type_live.data, playlist_type_live.len) == 0)
+	else
 	{
+		if (params[MEDIA_SET_PARAM_PLAYLIST_TYPE] != NULL &&
```
Comment (erankor) [inline vod/media_set_parser.c:2491]: no need to check params[MEDIA_SET_PARAM_PLAYLIST_TYPE] != NULL here, if we entered the 'else', it's surely not NULL

File: vod/media_set_parser.c [hunk 10]
```diff
@@ -2485,9 +2486,21 @@ media_set_parse_json(
...
+			params[MEDIA_SET_PARAM_PLAYLIST_TYPE]->v.str.len == playlist_type_event.len &&
+			vod_strncasecmp(params[MEDIA_SET_PARAM_PLAYLIST_TYPE]->v.str.data, playlist_type_event.data, playlist_type_event.len) == 0)
+		{
+			result->is_live_event = TRUE;
+		} else if (params[MEDIA_SET_PARAM_PLAYLIST_TYPE]->v.str.len != playlist_type_live.len ||
```
Comment (erankor) [inline vod/media_set_parser.c:2496]: please add newline between the } and else, that is the convention here

File: vod/media_set_parser.c [hunk 11]
```diff
@@ -2865,16 +2871,20 @@ media_set_parse_json(
 				parse_all_clips = FALSE;
 			}
 
+			// applying live window with infinite duration is necessary for playlistType event
+			// because we aren't allowed to remove segments from the head for this playlist type
```
Comment (erankor) [inline vod/media_set_parser.c:2875]: move the comment a bit down, near if (!result->is_live_event)

**Review by erankor — CHANGES_REQUESTED** [review]

**Comment by rhummelmose** [comment]
Thanks @eranko, changes as requested :)

**Comment by erankor** [comment]
Merged, thanks!

**Comment by rhummelmose** [comment]
Epic!

**Pull request merged**